packages/backend/src/services/UnfurlService/UnfurlService.ts

```diff
@@ -0,0 +1,332 @@
+import { ShareModel } from '../../models/ShareModel';
+import { SpaceModel } from '../../models/SpaceModel';
+import { getAuthenticationToken } from '../../routers/headlessBrowser';
+import { EncryptionService } from '../EncryptionService/EncryptionService';
+
+const puppeteer = require('puppeteer');
+
+const uuid = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}';
+const uuidRegex = new RegExp(uuid, 'g');
+const nanoid = '[\\w-]{21}';
+const nanoidRegex = new RegExp(nanoid);
+
+export enum LightdashPage {
+    DASHBOARD = 'dashboard',
+    CHART = 'chart',
+    EXPLORE = 'explore',
+}
+
+export type Unfurl = {
+    title: string;
+    description?: string;
+    imageUrl: string | undefined;
+    pageType: LightdashPage;
+};
+
+export type ParsedUrl = {
+    isValid: boolean;
+    lightdashPage?: LightdashPage;
+    url: string;
+    dashboardUuid?: string;
+    projectUuid?: string;
+    chartUuid?: string;
+    exploreModel?: string;
+};
+
+type UnfurlServiceDependencies = {
+    lightdashConfig: LightdashConfig;
+    dashboardModel: DashboardModel;
+    savedChartModel: SavedChartModel;
+    spaceModel: SpaceModel;
+    shareModel: ShareModel;
+    encryptionService: EncryptionService;
+};
+
+const saveScreenshot = async (
+    imageId: string,
+    cookie: string,
+    parsedUrl: ParsedUrl,
+): Promise<Buffer | undefined> => {
+    let browser;
+
+    try {
+        const browserWSEndpoint = `ws://${process.env.HEADLESS_BROWSER_HOST}:${process.env.HEADLESS_BROWSER_PORT}`;
+        browser = await puppeteer.connect({
+            browserWSEndpoint,
+        });
+
+        const page = await browser.newPage();
+
+        await page.setExtraHTTPHeaders({ cookie });
+
+        await page.setViewport({
+            width: 1400,
+            height: 768, // hardcoded
+        });
+
+        const blockedUrls = [
+            'headwayapp.co',
+            'rudderlabs.com',
+            'analytics.lightdash.com',
+            'cohere.so',
+            'intercom.io',
+        ];
+        await page.setRequestInterception(true);
+        page.on('request', (request: any) => {
+            const requestUrl = request.url();
+            if (blockedUrls.includes(requestUrl)) {
+                request.abort();
+                return;
+            }
+
+            request.continue();
+        });
+        await page.goto(parsedUrl.url, {
+            timeout: 100000,
+            waitUntil: 'networkidle0',
+        });
+        const path = `/tmp/${imageId}.png`;
+
+        const selector =
+            parsedUrl.lightdashPage === LightdashPage.DASHBOARD
+                ? '.react-grid-layout'
+                : `.echarts-for-react, [data-testid="visualization"]`; // Get .echarts-for-react, otherwise fallsback to data-testid (tables and bignumbers)
+        await page.waitForSelector(selector);
+        const element = await page.$(selector);
+
+        if (parsedUrl.lightdashPage === LightdashPage.DASHBOARD) {
```

Comment:
does this mean that it only works for dashboards now?

Reply:
this means that on dashboard we need some extra steps, we don't need to do this for explore or charts